Sources/SparseSetModule/SparseSet.swift

```diff
@@ -0,0 +1,643 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Collections open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+public struct SparseSet<Key, Value> where Key: FixedWidthInteger, Key.Stride == Int {
+  @usableFromInline
+  internal var _dense: DenseStorage
+
+  @inlinable
+  @inline(__always)
+  internal var _sparse: SparseStorage {
+    get {
+      SparseStorage(__sparseBuffer)
+    }
+    set {
+      __sparseBuffer = newValue._buffer
+    }
+  }
+
+  @usableFromInline
+  internal var __sparseBuffer: SparseStorage.Buffer
+}
+
+// MARK: -
+
+extension SparseSet {
+  /// A read-only collection view for the keys contained in this sparse set, as
+  /// a `ContiguousArray`.
+  ///
+  /// - Complexity: O(1)
+  @inlinable
+  @inline(__always)
+  public var keys: ContiguousArray<Key> {
+    _dense._keys
+  }
+
+  /// A mutable collection view containing the values in this sparse set.
+  ///
+  /// - Complexity: O(1)
+  @inlinable
+  @inline(__always)
+  public var values: Values {
+    get {
+      Values(_base: self)
+    }
+    _modify {
+      var values = Values(_base: self)
+      self = SparseSet()
+      defer { self = values._base }
+      yield &values
+    }
+  }
+}
+
+// MARK: -
+
+extension SparseSet {
+  /// The size of the key universe. The sparse set has enough space allocated to
+  /// store all (non-negative) keys less than this value. Adding a key to the
+  /// sparse set that is greater than or equal to the universe size will result
+  /// in extra memory allocation.
+  ///
+  /// - Complexity: O(1)
+  @inlinable
+  @inline(__always)
+  public var universeSize: Int { _sparse.capacity }
```

Comment:
Universe size seems like a completely different thing than capacity. In LLVM's implementation, universe size is actually a template parameter and is the static size of the sparse key array. The dense value vector is dynamically sized and I would usually think of capacity in terms of that.

Reply:
Like @milseman says we have two parameters: `universeSize` which is the size of the sparse storage and `capacity` which I think of as the the size of the dense storage.

In the current implementation we have an initializer `init(minimumCapacity: Int, universeSize: Int? = nil)` where we can specify both. If adding a key greater than or equal to `universeSize` then the sparse storage is resized to accommodate it. Dense storage is backed by `ContiguousArray` so will increase on size dynamically as needed.  `universeSize` is just a read-only property but it might be useful to add the ability to change the universe size manually after creation. 

At the moment there is no `capacity` property or `reserveCapacity(_ minimumCapacity: Int)` function but maybe we should add those and / or something like `reserveCapacity(_ minimumCapacity: Int, minimumUniverseSize: Int)`.

In any case we should mention the difference in the documentation as its a possible source of confusion.

Reply:
OK, if y'all think `universeSize` is the right term, then let's go with that!

I don't think we'd want to expose the dense array's capacity as a public property -- it exposes unstable information about allocation behavior, and I think making it public in the stdlib was a mistake that I'd prefer not to emulate in this package.

There is no such issue with `reserveCapacity`; adding it would probably be a good idea!